react-card: Spec (#19410)
microsoft/fluentui: packages/react-card/Spec.md

# @fluentui/react-card Spec

## Background

A cards main function is to provide the scaffolding for hosting actions and content for a single topic within a card sized object. It is a framework for organizing content within the confines of a card.

The card itself should not have content or actions built in, but provide the mechanisms for it to be displayed.

### Fabric (v8)

> 

Reviewer: There's actually a better previous attempt at a `Card` component in v8 under the [`@fluentui/react-cards`](https://github.com/microsoft/fluentui/tree/master/packages/react-cards) package.

Author: I think we can disregard this implementation as it's not publicly used. Unless you mean that this implementation is a better attempt than the `react-card` implementation we had?

Reviewer: No, I meant the former, but you are right, it's not publicly used so let's keep it off this file.

| Wrapper to render in column format                                                                                         |                      | CardColumn        | ❌        |
| Allows for content to expand on hover                                                                                      |                      | CardExpandableBox | ❌        |

...

## API Proposal

### Components

| Component   | Purpose                                                                  |
| ----------- | ------------------------------------------------------------------------ |
| Card        | Wrapper for the main content of a card.                                  |
| CardHeader  | Optional header slot for the card.                                       |
| CardBody    | Main slot of the Card                                                    |
| CardPreview | Previewing the document/article presented in the card through an image   |
| CardFooter  | Optional footer slot for the card. (non-applicable for horizontal cards) |

> 

Reviewer 2: can you please add a section for cards container? It does not necessarily have to be a separate component, but we should at least have documentation about how to group several cards and what interactions are possible.

Cards are usually laid out in a grid, so we need to decide how should we represent that grid from accessibility point of view. It might be challenging as there are no good roles that would represent this - `menu` and `list`/`listbox` are linear, `grid` has strict DOM structure requirements that usually does not play well with layout and also brings narration of rows/columns that might be confusing. 

As for navigation between the cards, we should look into tabster support for 2 dimensional grids and the ability to enter/escape to/from card. Alternative would be to use tab to go inside of the card. 

Probably somehing to discuss separately, but since cards are in a container most of the time, it is an important topic.

Author: I understand the problem but think it should definitely be covered elsewhere. Adding that would add to the complexity of this Spec as it's now also covering layout scenarios of cards.
While important, I think it should either be a component of itself or documented/defined elsewhere as the focus is different from a Card implementation.
What do you think?

### Card

#### Anatomy

![visual anatomy of the Card component](./assets/Card.png)

#### API

| Property    | Values                                                                               | Default    | Purpose                                                                                   |
| ----------- | ------------------------------------------------------------------------------------ | ---------- | ----------------------------------------------------------------------------------------- |
| orientation | `vertical`, `horizontal`                                                             | `vertical` | Orientation of the card                                                                   |

> | size        | `smallest`, `smaller`, `small`, `medium`, `large`                                    | `medium`   | Define the minimum size of the card. Smaller sizes only apply to horizontal card          |

Reviewer: Shouldn't we also have `larger` and `largest` if we have `smaller` and `smallest` as possible sizes?

Author: Design does not differentiate between `large`, `larger` and `largest`. Having them would have no implementation difference.

| scale       | `fixed`, `auto-width`, `auto-height`, `auto`, `fluid-width`, `fluid-height`, `fluid` | `auto`     | Manages how the card handles it's scaling depending on the content                        |
| onClick     | function                                                                             | undefined  | Function to call when clicking the card. Makes the card assume the interactive appearance |

> | appearance  | `filled`, `filled-alternative`, `outline`, `subtle`                                  | `filled`   | Define the appearance of the card                                                         |

Reviewer: Not a complaint on this spec itself, but we should really define what the pattern for this is going forward as it is all over the place (i.e. one prop with defined values here vs a prop for each in `Button`).

Author: That's a good point. There's no previous defined pattern for this right? I only went with this approach as it seems to me to be much more manageable than several different props.
Are you aware of other components with appearance variations apart from `Button`?

Reviewer: I think there are a couple other components that use it, but I don't remember which specific ones.

Reviewer 3: I think this [RFC](19775) suggests using one prop with defined values.

| selectable  | boolean                                                                              | false      | Makes the card selectable by adding a checkbox to the _Actions_ area                      |
| selected    | boolean                                                                              | false      | Set to `true` if card is selected                                                         |
| expandable  | boolean                                                                              | false      | Allow card to expand to show whole content                                                |

> | context     | _slot_                                                                               | undefined  | Slot for the context menu, when right-clicking the card                                   |

Reviewer 2: the word context might be a bit confusing - it usually suggest the React context, not the onContextMenu. Maybe we can rename this to `contextMenu`. Also, you plan to use the same slot for the ... (three dots) menu? Because that might be slightly different from the menu that appears on right click. Lastly, do we need any change in the Menu compoennt or MenuList so that it can be integrated in the slot?

Author: That's a good point, I will rename this to `contextMenu` as suggested.
For the three dots, there isn't enough information on the design spec that clarifies this, as only the horizontal card has it. As such, my thoughts were that that it's up to the user to add any additional menu they want and that this only covers the right-click/long press scenarios.
Regarding `Menu`/`MenuList`, I am unaware. As long as it enables us to have a menu in one place and apply the trigger wherever, all is good.

| disabled    | boolean                                                                              | false      | Makes the card disabled                                                                   |

#### `scale` property

- `fixed`: Size of the card is fixed.
- `auto-width`: `width` is set to `fit-content`.
- `auto-height`: `height` is set to `fit-content`.
- `auto`: `width` and `height` are set to `fit-content`.
- `fluid-width`: `width` is set to `100%`.
- `fluid-height`: `height` is set to `100%`.
- `fluid`: `width` and `height` are set to `100%`.

### CardHeader

#### Anatomy

![visual anatomy of the CardHeader component](./assets/CardHeader.png)

#### API

| Property    | Values | Default   | Purpose                                           |
| ----------- | ------ | --------- | ------------------------------------------------- |
| image       | _slot_ | undefined | Logo or image to better contextualize the content |

> | title       | _slot_ | undefined | Main title of the card                            |

Reviewer 2: `title` collides with HTML `title` attribute name, so most probably we should find a different name, for example header

Author: Good point, changed to `header`

## Sample Code

### Compact

```jsx
<Card orientation="horizontal">
  <CardHeader image={<img src="powerpoint.png" />} title="Team offsite 2020" description="Onedrive > Files" />
</Card>
```

### App

```jsx
<Card>
  <CardHeader image={<img src="app_logo.png" />} title="App Name" description="Developer" />
  <CardBody>
    Donut chocolate bar oat cake. Dragée tiramisu lollipop bear claw. Marshmallow pastry jujubes toffee sugar plum.
  </CardBody>
</Card>
```

### Task

```jsx
<Card>
  <CardHeader title="Alert in Teams when a new document is uploaded in channel" description="By Microsoft" />
  <CardFooter>
    <span>Automated</span>
    <span>3290 uses</span>
  </CardFooter>
</Card>
```

### Image

```jsx
<Card>
  <CardPreview src="monkey.png" alt="image of monkey" />
  <CardBody>
    <div>
      <span>Length</span>
      <span>4.6-39.6 IN</span>
      <span>11.7-100.6 cm</span>
    </div>
    <div>
      <span>Diet</span>
      <span>Omnivore</span>
      <span>nuts, fruits, eggs</span>
    </div>
  </CardBody>
  <CardFooter>
    <Button>Share</Button>
  </CardFooter>
</Card>

> ```

Reviewer: What do we plan on doing if someone passes children other than `CardPreview`, `CardHeader`, `CardBody` or `CardFooter` to the `Card`?

Reviewer 2: I think other children are allowed and expected, as CardPreview, CardHeader, CardBody or CardFooter for example might be wrapped by a data fetching component

Author: The idea is to have the inner components apply any self styling so it's up to the user to arrange the components as intended. This opens up room for the user to put a CardPreview on top with a CardHeader under it (or vice-versa) and for you to create your own special card components, giving you the freedom to customize.
Another option would be to have slots in the card for Header and Footer and have everything else just render in the middle, restricting usage.
Thoughts?

Reviewer: I think freedom to customize makes sense, you can disregard my original comment.

## Behaviours

### Mouse

#### Selectable cards

![Visual interaction of selectable cards using the mouse](./assets/selectable-interaction-mouse.png)

Clicking the card surface can toggle the checkbox state to `selected`. Note that checkboxes are composed together as an additional component and are not provided by default.

1. Rest Default state is `unselected`
2. Hover cursor style changes to `pointer`, background color and elevation follows interactive styling to show visual feedback
3. Pressed cursor does not change background color and elevation follows interactive styling to show visual feedback for pressed state.

> 4. MouseUp Card and Checkbox are both in selected state, clicking again toggles components to unselected state

Reviewer 2: I think we should use click handler for this interaction, not mouse up. I am not sure if screen readers send mouse up/down events in all modes.

Reviewer: Agreed, if we are really adamant in using up/down events we should use pointer events instead of mouse events but listening to the click event should be enough.

Author: Agreed, changed to click

### Non-interactive cards

![Visual interaction of non-interactive cards using the mouse](./assets/non-interactive-interaction-mouse.png)

> Cards that have multiple actionable child components within them cannot be clickable. We preserve the use of `EnterKey` for moving focus to actionable controls inside the card, therefore the card surface cannot be used to set the state of the card but is reliant on the checkbox or similar inputs to set the card state.

Reviewer: > Cards that have multiple actionable child components within them cannot be clickable.

Are we planning to enforce this via code? If so, how?

Author: That's something that I've been thinking might not be feasible after starting implementation. An option would be to disable `pointer-events`, but this does not disable keyboard interactions with the elements. Plus, styling unknown components is a tough challenge.
Only thing that could be feasible would be to style the card and any button elements under it as disabled. Unsure how to tackle the event blocking. Any ideas?

Reviewer: It is definitely a very cumbersome and non-feasible problem to tackle imo, which is why I made the comment in the first place. Maybe the best way is to call out that expectation in the spec and indicate that if the user insists on having actionable child components within a clickable card then it's up to them to manage the bubbling of events.

Something like

```suggestion
Cards that have multiple actionable child components are expected to not be clickable. We preserve the use of `EnterKey` for moving focus to actionable controls inside the card, therefore the card surface should not be used to set the state of the card but is reliant on the checkbox or similar inputs to set the card state.

If users have a scenario that requires them to use actionable components within a clickable Card then the bubbling of events is expected to be managed by the users themselves.
```

Author: I'll make the changes to cover this and add your suggestion, I really don't think it's feasible to have Card manage child components' disabled state effectively.

Mouse interaction

- OnMouseEnter / Hover cursor style stays at default, rest state stays persistent no visual feedback is applied to the card surface
- Hover & onClick (Checkbox) Hover state & pressed state applies to checkbox component

> - MouseUp (Checkbox) Card and Checkbox are both in selected state, clicking again toggles components to unselected state

Reviewer 2: I think we should use click handler for this interaction, not mouse up. I am not sure if screen readers send mouse up/down events in all modes.

Author: Agreed, fixed

### Context menu

![Visual interaction of the context menu on cards using the mouse](./assets/context-interaction-mouse.png)

A context menu can be triggered by right clicking the surface or through an action in the card.

> 1.Rest

Reviewer: nit:
```suggestion
1. Rest
```

2. RightClick [LongPress]
   Right click anywhere on card surface invokes content menu
3. Show
   Context menu appears close to cursor location on card
4. Hide/Dismiss
   Clicking or tapping anywhere outside of menu to dismiss

#### Disabled

![Visual interaction of disabled cards using the mouse](./assets/disabled.png)

1 Unselected disabled

1. Label: Selectable Card
2. Role: Group
3. State: Disabled, unselected

2 Selected disabled

1. Label: Selectable Card
2. Role: Group
3. State: Disabled, selected

### Keyboard

#### Selectable cards

![Visual interaction of selectable cards using the keyboard](./assets/selectable-interaction-keyboard.png)

1. Rest Default state is ‘unselected’
2. Focus Focus rectangle is shown on card
3. Focus Use spacebar to toggle state (selected/unselected)

> ### Non-interactive cards

Reviewer 2: The heading says non-interactive, but below you describe interaction with focus inside. Is it wrong heading?

Reviewer: I think it means that the Cards themselves are not interactive but the contents can be. Maybe non-clickable Cards would be a better heading?

Author: What @khmakoto said. Non-interactive cards are cards that do not have an action when being clicked on (no onClick defined), think for example a Microsoft Store scenario where you have apps in cards and clicking on them would open a new page.
What do you mean by focus inside?

![Visual interaction of non-interactive cards using the keyboard](./assets/non-interactive-interaction-keyboard.png)

Keyboard interaction

1. Focus Initial focus is set on card, to access items within the card use EnterKey or Spacebar
2. Focus-Inside Focus moves to first actionable item in the card. Use ArrowKeys or TabKey to move focus to child elements within the card. To exit card use EscKey, to move focus back to card [parent]
3. Checkbox-Focus Depending on your tab order of items within the card, the number of keystrokes will vary. Once focus is set on the checkbox by default it will be set to unselected

> 4. SpaceBar or EnterKey will toggle both Card and Checkbox to selected state. Keying again will toggle to unselected

Reviewer 2: in 1. it is written that enter and space move focus to the card, but in 4 that they select the card. Did you mean to specify 4 only for when checkbox is focused?

Reviewer: I think that is the case since it comes after you set focus on the Checkbox in 3, but I agree it is a bit confusing in its wording, maybe it could be something like 
```suggestion
4. SpaceBar or EnterKey while focused on the Checkbox will toggle both Card and Checkbox to selected state. Keying again will toggle to unselected
```

Author: This describes a user interaction flow. Given that it is mentioned we're focusing the checkbox on point 3., pressing Space/Enter on 4. has the pre-existing condition that we will be focused on the Checkbox.
I will add @khmakoto's suggestion regardless for clarity purposes.

> 

Reviewer 2: I think there is one more scenario for keyboard/screen reader and mouse/touch - clickable (focusable) cards. These are not selectable (do not have selected/unselected state) and do not include other focusable elements inside of them, but the user can still click on the card.

In this scenario, you need to make the card focusable, but enter/space would not move focus inside if the card - instead it would call the click handler (unless the card is disabled)

Author: Added, thanks!

### Context menu

![Visual interaction of the context menu on cards using the keyboard](./assets/context-interaction-keyboard.png)

1. Rest

> 2. OnFocus - Focus rect. is set on card. Use shiftKey and press F10 to invoke context menu

Reviewer: Should Alt+Down also invoke the context menu?

Reviewer 2: That would not be expected, maybe alt+down could expand the collapsed card as that is the shortcut for opening dropdowns.

Author: I agree with @jurokapsiar, `Alt + Down arrow` could expand expandable cards.

See if you agree with this behavior or if there's anything else to be added to it:
1. Focus expandable card
2. Press `Alt + Down arrow` to expand the card

>    `aria-haspopup="menu"`

Reviewer 2: aria-haspopup suggest that a menu would appear on (left) click so it is not a good fit for indicating a context menu. In Teams, the pattern is to add a dummy element with "has context menu" text and use it in aria-describedby. Not ideal so open to other suggestions, but so far this was the best we could find.

Reviewer: CCing @smhigley for discussion about this in particular

   `aria-expanded="false"`

3. ShowSet focus on context menu

   `aria-expanded="true"`

4. Hide/Dismiss - Use EscKey to hide/dismiss

   `aria-expanded="false"`

## Accessibility

> ⚠️ Note: All other accessibility information, not covered in this section, is provided throughout the spec.

Each card is of `role="group"`.

### Narrator guidance

#### Selectable cards

![Visual interaction of keyboard usage with selectable cards](./assets/selectable-narrator.png)

1 Focus + unselected

1. Label: Select this card
2. Role: Group
3. State: Focus + unselected

> 4. Helper text: Use enterKey or spacebar to toggle state to selected

Reviewer 2: there is no good way to add such custom helper text to screen reader narration. You can consider:
- aria-describedby linking to an element with the description text
- aria live narration when (first) card gets focused.
Both have pros and cons, I would suggest to set up a meeting with @kolaps33 and @smhigley to discuss pros and cons or alternatives.

Author: I suggest the discussion remains here so it's documented what the pros and cons are, to keep our reasoning for why we choose option A or B.

2 Focus + selected

1. Label: Unselect this card
2. Role: Group
3. State: Focus + selected
4. Helper text: Use enterKey or spacebar to toggle state to unselected

#### Non-interactive cards

![Visual interaction of keyboard usage with non-interactive cards](./assets/non-interactive-narrator.png)

1 Focus + unselected

1. Label: App Card
2. Role: Group
3. State: Focus
4. Helper text: Use enterKey to access controls within the card

2 Focus (inside)

1. Label: Open App
2. Role: Button
3. State: Focus
4. Helper text: Use enterkey to open app in new window

3 Focus Checkbox unselected

1. Label: Share App
2. Role: Button
3. State: Focus
4. Helper text: Use enter key to open share dialog

4 Focus Checkbox selected

1. Label: Select card
2. Role: Checkbox
3. State: Unselected
4. Helper text: Use spacebar or enter key to select this item

#### Disabled

> ![Two disabled cards. One is unselected and one is selected](./assets/disabled-narrator.png)

Reviewer: Are we implicitly saying that we will allow focus on disabled Cards? I don't think this is something we should allow.

Author: The point here is to portray disabled cards in different states. A selected card **is not** a focused card.
I can see this behavior useful in scenarios where you have cards selected but you're in a read-only mode. Your card is selected, but disabled as you can't change it in read-only.

Reviewer: My bad, I got confused by the borders around the Cards but I see now that they were only there to emphasize them, please disregard this comment.

1 Unselected disabled

1. Label: Selectable Card
2. Role: Group
3. State: Disabled, unselected

2 Selected disabled

1. Label: Selectable Card
2. Role: Group
3. State: Disabled, selected

#### Context menu

![Context menu opening and closing](./assets/context-narrator.png)

RightClick / ShiftKey + F10

1. OnFocus

   `aria-haspopup="true"`

>    `aria-expanded="false"`

Reviewer 2: would aria-expanded be used for context menu? Wasn't it used for expanded/collapsed cards?

Author: Is `aria-expanded` relevant for expandable cards? The content is there, it's just not visible until hovered.
There's also no concept of collapsible cards.

>    `aria-controls="menu"`

Reviewer 2: usually menu appears in DOM only after it is triggered, so it is not possible to reference it in aria-controls before opening.

Author: Is there a suggestion or should this just be removed?

| disabled    | boolean                                                                              | false      | Makes the card and card selection disabled (not propagated to children)                   |

#### `scale` property

- `fixed`: Size of the card is fixed.
- `auto-width`: `width` is set to `fit-content`.
- `auto-height`: `height` is set to `fit-content`.
- `auto`: `width` and `height` are set to `fit-content`.
- `fluid-width`: `width` is set to `100%`.
- `fluid-height`: `height` is set to `100%`.
- `fluid`: `width` and `height` are set to `100%`.

#### Context menu

Context menu should be handled in conjunction with the @fluentui/react-menu component. A simple usage example would be:

```jsx
<Menu openOnContext>
  <MenuTrigger>
    <Card />
  </MenuTrigger>

  <MenuPopover>
    <MenuList>
      <MenuItem>Share</MenuItem>
      <MenuItem>Dismiss</MenuItem>
    </MenuList>
  </MenuPopover>
</Menu>
```

`react-card` documentation will contain more elabora examples to cover the different use cases.

> ### CardHeader

Reviewer 4: DOM output would be great for each of these sub-components

### Task

```jsx
<Card>
  <CardHeader header="Alert in Teams when a new document is uploaded in channel" description="By Microsoft" />
  <CardFooter>
    <span>Automated</span>
    <span>3290 uses</span>
  </CardFooter>
</Card>
```

### Image

```jsx
<Card>
  <CardPreview src="monkey.png" alt="image of monkey" />
  <CardBody>
    <div>
      <span>Length</span>
      <span>4.6-39.6 IN</span>
      <span>11.7-100.6 cm</span>
    </div>
    <div>
      <span>Diet</span>
      <span>Omnivore</span>
      <span>nuts, fruits, eggs</span>
    </div>
  </CardBody>
  <CardFooter>
    <Button>Share</Button>
  </CardFooter>
</Card>
```

## Behaviours

### Mouse

### Interactive cards

![Visual interaction of interactive cards using the mouse](./assets/interactive-interaction-mouse.png)

The card surface can be used as a clickable target

1. Rest

2. OnHover

   Cursor style changes to pointer, background color & elevation values change to show card is an actionable element

3. OnClick

   Executes action assigned to card i.e. “Open file, Open link in browser etc...”

#### Selectable cards

![Visual interaction of selectable cards using the mouse](./assets/selectable-interaction-mouse.png)

> Clicking the card surface can toggle the checkbox state to `selected`. Note that checkboxes are composed together as an additional component and are not provided by default.

Reviewer 4: > checkboxes are composed together as an additional component and are not provided by default

What does this mean ? can you add some sample code for this scenario ?

Author: This was leftover from the design spec, I'll remove it. Context here is that the Checkbox is not Card implementation but instead it's own component, I think.

3. Pressed

   Cursor does not change background color and elevation follows interactive styling to show visual feedback for pressed state.

4. onClick

   Card and Checkbox are both in selected state, clicking again toggles components to unselected state

### Non-interactive cards

![Visual interaction of non-interactive cards using the mouse](./assets/non-interactive-interaction-mouse.png)

Cards that have multiple actionable child components are expected to not be clickable. We preserve the use of `EnterKey` for moving focus to actionable controls inside the card, therefore the card surface should not be used to set the state of the card but is reliant on the checkbox or similar inputs to set the card state.

If users have a scenario that requires them to use actionable components within a clickable Card then the bubbling of events is expected to be managed by the users themselves.

Mouse interaction

- OnMouseEnter / Hover

  Cursor style stays at default, rest state stays persistent no visual feedback is applied to the card surface

- Hover & onClick (Checkbox)

  Hover state & pressed state applies to checkbox component

> - onClick

Reviewer 4: Does this `onClick` apply to the entire card ? wouldn't that just turn the card interactive ?

>   (Checkbox) Card and Checkbox are both in selected state, clicking again toggles components to unselected state

Reviewer 4: Can you please clarify in the spec when the card is a tabstop when it has a checkbox and is (not) interactive

> #### Disabled

Reviewer 4: Should an interactive card still be tabstop when disabled ? or do we need to support a `disabledFocusable` prop for the component ?

Author: It should not. I've added a section for Disabled in the interactions stating that a card should not receive focus when disabled.

![Visual interaction of Disabled cards using the mouse](./assets/disabled.png)

1. Unselected disabled

   - Label: Selectable Card
   - Role: Group
   - State: Disabled, unselected

2. Selected disabled

   - Label: Selectable Card
   - Role: Group
   - State: Disabled, selected

### Keyboard

### Interactive cards

![Visual interaction of interactive cards using the keyboard](./assets/interactive-interaction-keyboard.png)

1. Rest

2. Keyboard interaction

   EnterKey Perfoms action on card “Open file, Open link in browser etc...”

#### Selectable cards

![Visual interaction of selectable cards using the keyboard](./assets/selectable-interaction-keyboard.png)

1. Rest Default state is ‘unselected’
2. Focus Focus rectangle is shown on card
3. Focus Use spacebar to toggle state (selected/unselected)

### Non-interactive cards

![Visual interaction of non-interactive cards using the keyboard](./assets/non-interactive-interaction-keyboard.png)

Keyboard interaction

1. Focus

   Initial focus is set on card, to access items within the card use EnterKey or Spacebar

> 2. Focus-Inside

Reviewer 4: Can you please specify how to focus out of the card ? is it just a simple `Tab` press or is it expected to trap focus inside the card until `Escape` is pressed

   Focus moves to first actionable item in the card. Use ArrowKeys or TabKey to move focus to child elements within the card. To exit card use EscKey, to move focus back to card [parent]

3. Checkbox-Focus

>    Depending on your tab order of items within the card, the number of keystrokes will vary. Once focus is set on the checkbox by default it will be set to unselected

Reviewer 4: Does this mean that if I focus a selected checkbox it will be unselected ? 👀

Author: Poor phrasing in the design spec, supposed to say that by default the checkbox is unselected and not that it will be unselected when focusing. Removed that altogether.

4. SpaceBar or EnterKey, while focused on the Checkbox, will toggle both Card and Checkbox to selected state. Keying again will toggle to unselected

## Accessibility

> ⚠️ Note: All other accessibility information, not covered in this section, is provided throughout the spec.

Each card is of `role="group"`.

### Narrator guidance

### Interactive cards

![A card with the number 1 over it](./assets/interactive-narrator.png)

1. Label: App Card
2. Role: Group
3. State: Focus
4. Helper text: Use enterKey to perform action

#### Selectable cards

![Visual interaction of keyboard usage with selectable cards](./assets/selectable-narrator.png)

> 1 Focus + unselected

Reviewer 4: It would be great to have some DOM output in the spec somewhere to make these points clearer

Author: Let me know if the DOM I added above for each Card component is enough or if I should also put it here.

# @fluentui/react-card Spec

## Background

> A cards main function is to provide the scaffolding for hosting actions and content for a single topic within a card sized object. It is a framework for organizing content within the confines of a card.

Reviewer: nit:
```suggestion
A card's main function is to provide the scaffolding for hosting actions and content for a single topic within a card sized object. It is a framework for organizing content within the confines of a card.
```

| onClick     | function                                                                             | undefined  | Function to call when clicking the card. Makes the card assume the interactive appearance |
| appearance  | `filled`, `filled-alternative`, `outline`, `subtle`                                  | `filled`   | Define the appearance of the card                                                         |
| selectable  | boolean                                                                              | false      | Makes the card selectable by adding a checkbox to the _Actions_ area                      |
| selected    | boolean                                                                              | false      | Set to `true` if card is selected                                                         |
| expandable  | boolean                                                                              | false      | Allow card to expand to show whole content                                                |

> | contextMenu | _slot_                                                                               | undefined  | Slot for the context menu, when right-clicking the card                                   |

Reviewer 4: I do not think that a contextMenu should be in a `Card` as a slot. The current `Menu` implementation encourages composition of components through JSX. What is wrong with this kind of API to support a context menu scenario for card ? 

https://codesandbox.io/s/optimistic-cache-975ji?file=/src/App.js

Reviewer 4: Here is another example if we don't want embed the `Card` inside a `MenuTrigger` for a context menu: https://codesandbox.io/s/falling-tree-oh6o0?file=/src/App.js

In order to support a context menu slot we would also need to bundle `react-menu` somehow with it adding to bloat.

I would say that we should look for some use case that we cannot achieve with the existing APIs that we have and then see what is the best way to achieve them. Menu slots in v0/8 are generally very complicated an hard to memoize/virtualize since they are just js objects. The 2 methods I've proposed here all leverage plain JSX.

Author: Yeah this makes much more sense and option 2 also allows us to reuse the menu for when you click an ellipsis menu scenario. Thanks for the example @ling1726, do you think this should be covered in the spec or docs (or both)?

Reviewer 4: This should be covered in the docs IMO.

Author: Awesome, then I'll just add a section to explain this should be covered with Menu instead and then in the future use your examples in the docs. Thanks!

Cards that have multiple actions within them cannot be clickable. We preserve the use of `EnterKey` for moving focus to actionable controls inside the card.

1. OnMouseEnter / Hover

   Cursor style stays at default, rest state stays persistent no visual feedback is applied to the card surface

2. Hover & onClick

   Hover state only applies to actionable items within card i.e. buttons, links etc. Cursor changes to pointer (or follows behavior of that component) ”

#### Selectable cards

![Visual interaction of selectable cards using the mouse](./assets/selectable-interaction-mouse.png)

Clicking the card surface can toggle the checkbox state to `selected`. Note that checkboxes are composed together as an additional component and are not provided by default.

1. Rest

   Default state is `unselected`

2. Hover

   Cursor style changes to `pointer`, background color and elevation follows interactive styling to show visual feedback

3. Pressed

   Cursor does not change background color and elevation follows interactive styling to show visual feedback for pressed state.

4. onClick

>    Card and Checkbox are both in selected state, clicking again toggles components to unselected state

Reviewer 3: ```suggestion
   Card and Checkbox are both in selected state, clicking again toggles components to unselected state.
```

If users have a scenario that requires them to use actionable components within a clickable Card then the bubbling of events is expected to be managed by the users themselves.

- OnMouseEnter / Hover

  Cursor style stays at default, rest state stays persistent no visual feedback is applied to the card surface

- Hover & onClick (Checkbox)

  Hover state & pressed state applies to checkbox component

- onClick (Checkbox)

  Card and Checkbox are both in selected state, clicking again toggles components to unselected state

#### Disabled

![Visual interaction of Disabled cards using the mouse](./assets/disabled.png)

In scenarios where a selectable card may be disabled both the card and the checkbox should be set to disabled.
Hovering the card changes the cursor style to not-allowed

1. Unselected disabled

   - Label: Selectable Card
   - Role: Group
   - State: Disabled, unselected

2. Selected disabled

   - Label: Selectable Card
   - Role: Group
   - State: Disabled, selected

### Keyboard

### Interactive cards

![Visual interaction of interactive cards using the keyboard](./assets/interactive-interaction-keyboard.png)

1. Rest

2. Keyboard interaction

>    `EnterKey` Perfoms action on card “Open file, Open link in browser etc...”

Reviewer 3: ```suggestion
   `EnterKey` Perfoms action on card “Open file, Open link in browser etc...”.
```

### Non-interactive cards with interactive child components

![Visual interaction of interactive cards using the keyboard](./assets/non-interactive-interaction-keyboard.png)

1. Focus

>    Initial focus is set on card, to access items within the card use `EnterKey` or `Spacebar`

Reviewer 3: ```suggestion
   Initial focus is set on card, to access items within the card use `EnterKey` or `Spacebar`.
```

2. Focus-Inside

>    Focus moves to first actionable item in the card. Use `ArrowKeys` or `TabKey` to move focus to child elements within the card. To exit card use `EscKey`, to move focus back to card

Reviewer 3: ```suggestion
   Focus moves to first actionable item in the card. Use `ArrowKeys` or `TabKey` to move focus to child elements within the card. To exit card use `EscKey`, to move focus back to card.
```

#### Selectable cards

![Visual interaction of selectable cards using the keyboard](./assets/selectable-interaction-keyboard.png)

1. Rest

   Default state is ‘unselected’

2. Focus

   Focus rectangle is shown on card

3. Focus

   Use spacebar to toggle state (selected/unselected)

### Non-interactive selectable cards

![Visual interaction of non-interactive selectable cards using the keyboard](./assets/non-interactive-selectable-interaction-keyboard.png)

1. Focus

   Initial focus is set on card, to access items within the card use `EnterKey` or `Spacebar`.

2. Focus-Inside

   Focus moves to first actionable item in the card. Use `ArrowKeys` or `TabKey` to move focus to child elements within the card. To exit card use `EscKey`, to move focus back to card.

3. Checkbox-Focus

   Depending on your tab order of items within the card, the number of keystrokes will vary.

4. `SpaceBar` or `EnterKey` while focused on the Checkbox

>    Will toggle both Card and Checkbox to selected state. Keying again will toggle to unselected

Reviewer 3: ```suggestion
   Will toggle both Card and Checkbox to selected state. Keying again will toggle to unselected.
```
